lock-impl/src/test/java/com/palantir/lock/impl/ThreadInfoLockServiceImplTest.java

```diff
@@ -0,0 +1,308 @@
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_2));
+
+        lockService.unlock(response2.getToken());
+        forceSnapshot();
+
+        // Only T3 should hold locks
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2))
+                .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, TEST_THREAD_3));
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3)).isNull();
+
+        lockService.unlock(response3.get().getToken());
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3)).isNull();
+    }
+
+    @Test
+    public void refreshDoesNotChangeSnapshot() throws InterruptedException {
+        LockResponse response =
+                lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+
+        lockService.refreshTokens(List.of(response.getToken()));
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+    }
+
+    @Test(timeout = 500L)
+    public void backgroundSnapshotRunnerWorks() throws InterruptedException {
+        LockServiceImpl lockServiceWithBackgroundRunner = LockServiceImpl.create(LockServerOptions.builder()
+                .isStandaloneServer(false)
+                .threadInfoConfiguration(ImmutableDebugThreadInfoConfiguration.builder()
+                        .recordThreadInfo(true)
+                        .threadInfoSnapshotIntervalMillis(10L)
+                        .build())
+                .build());
+        LockThreadInfoSnapshotManager backgroundSnapshotRunner = lockServiceWithBackgroundRunner.getSnapshotManager();
+
+        LockResponse response = lockServiceWithBackgroundRunner.lockWithFullLockResponse(
+                LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+
+        Thread.sleep(50L);
+
+        assertThat(backgroundSnapshotRunner.getLastKnownThreadInfoSnapshot().get(TEST_LOCK_1))
+                .isEqualTo(ANONYMOUS_TEST_THREAD_1);
+
+        lockServiceWithBackgroundRunner.unlock(response.getToken());
+
+        Thread.sleep(50L);
+
+        assertThat(backgroundSnapshotRunner.getLastKnownThreadInfoSnapshot().get(TEST_LOCK_1))
+                .isNull();
+    }
+
+    private void forceSnapshot() {
+        this.lockService.getSnapshotManager().takeSnapshot();
+    }
+
+    private LockClientAndThread getLatestThreadInfoForLock(LockDescriptor lock) {
+        return lockService.getSnapshotManager().getLastKnownThreadInfoSnapshot().get(lock);
```

Comment:
Prefer returning the map and using assertThat().hasEntry, _or_ returning an optional, instead of the current null checks.

Reply:
Went with optional